[Merged by Bors] - feat: miscellaneous improvements to the differential geometry elaborators

grunweg opened the pull request:
- Support inferring the model with corners on a `TangentBundle`.
- Support inferring the model with corners on the space `H` a manifold is modelled on.
- Correct the grammar in an error message.
- Extend the tracing to show a bit more information.
- Use backticks in all tracing (and other) error messages, per [zulip discussion](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Style.3A.20Backticks.20in.20messages.3F/with/544859886)
- Add some more tests, including one which is currently failing.

---

- [x] depends on: #30307

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
Two last tests for partial homeomorphisms and equivalences.

grunweg committed:
chore: move tests to a better place

grunweg committed:
More tests

grunweg committed:
feat: implement inferring the model with corners for the tangent bundle

thorimur reviewed (commented):
Looks good! I'm excited to get these elaborator improvements in at a steady pace! :D (I'd like to give it a once-over after #30307 lands just to make sure the diff is reasonable.)

thorimur committed:
Review comments

grunweg committed:
chore: add tests for charts and extended charts

The former test currently fails. Also add a failing test for the missing
handling of a ContinuousLinearEquiv type.

grunweg committed:
feat: infer the model with corners on the model space H of a manifold itself

grunweg committed:
chore: fix grammar of error message

grunweg committed:
chore: rename fromChartedSpace to fromManifold

This is what it does: it infers a model with corners on a manifold.
A future PR will also infer the same model on underlying model space of
the manifold, but in both cases, this relies on a manifold instance.

grunweg commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -302,15 +302,19 @@ where
+    -- Return an expression for a type `H` (if any) such that `e` is a ChartedSpace over `H`,
+    -- or `e` is `H` itself.
     let some H ← findSomeLocalInstanceOf? ``ChartedSpace fun _ type ↦ do
+        trace[Elab.DiffGeo.MDiff] "found a `ChartedSpace` instance: `{type}`"
```
It certainly was useful for me when debugging why my first version of the code failed. (I'm open to removing it again... not sure how much tracing output is desired.)

When we find a model, we report that one already --- but not the ChartedSpace instance... should we? What do you think? (I'm fine either way.)

grunweg committed:
Update Mathlib/Geometry/Manifold/Notation.lean

Co-authored-by: thorimur <68410468+thorimur@users.noreply.github.com>

grunweg committed:
Update

grunweg committed:
Better tracing for charted space search

grunweg committed:
Tweak other tracing in the same way

grunweg committed:
Fix typo

thorimur committed:
chore: apply suggestion

grunweg committed:
Tweak one message, move another

grunweg committed:
chore: put backticks around expressions consistently

Vierkantor reviewed (approved):
The meta code looks fine to me, and this looks really well tested! But this really is not my area of expertise, mathematics-wise, so I really cannot say if the tests cover the functionality well. Could a differential geometer confirm that this makes sense?

maintainer merge
